File: solutions/HJK9810/2026-08-25/high/5644.py
DIR = [(0, 0), (-1, 0), (0, 1), (1, 0), (0, -1)]

def sort_charge(case_num, charge_case):
    sort_case = [num for num in case_num]
    size = len(case_num)

    for idx in range(size - 1, -1, -1):
        for index in range(idx):
            if charge_case[sort_case[index]] < charge_case[sort_case[index + 1]]:
                sort_case[index + 1], sort_case[index] = sort_case[index], sort_case[index + 1]

    return sort_case

def move_pos(current, dir_move):
    row, col = current
    d_row, d_col = DIR[dir_move]
    return [row + d_row, col + d_col]

def initial_baord(num):
    charge_case = [0]
    boards = [[[0] for _ in range(10)] for _ in range(10)]

    for idx in range(num):
        col, row, size, charge = map(int, input().split())
        row -= 1
        col -= 1
        charge_case.append(charge)

        for r in range(row - size, row + size + 1):
            for c in range(col - size, col + size + 1):
                if 0 <= r < 10 and 0 <= c < 10 and abs(r - row) + abs(c - col) <= size:
                    boards[r][c].append(idx + 1)

    return boards, charge_case

def algorithm():
    M, A = map(int, input().split())
    A_move = list(map(int, input().split()))
    B_move = list(map(int, input().split()))

    boards, charge_case = initial_baord(A)
    total_charge = 0

    A_move.insert(0, 0)
    B_move.insert(0, 0)

    A_pos = [0, 0]
    B_pos = [9, 9]
    for move in range(M + 1):
        A_pos = move_pos(A_pos, A_move[move])
        B_pos = move_pos(B_pos, B_move[move])

        A_charge = sort_charge(boards[A_pos[0]][A_pos[1]], charge_case)
        B_charge = sort_charge(boards[B_pos[0]][B_pos[1]], charge_case)

        if len(A_charge) == 1 and len(B_charge) == 1: continue
        elif A_charge[0] == B_charge[0]:
            total_charge += (charge_case[A_charge[0]] + max(charge_case[A_charge[1]], charge_case[B_charge[1]]))
        else:            
            total_charge += (charge_case[A_charge[0]] + charge_case[B_charge[0]])

    return total_charge

T = int(input())
# 여러개의 테스트 케이스가 주어지므로, 각각을 처리합니다.
for test_case in range(1, T + 1):
    print(f"#{test_case} {algorithm()}")

File: solutions/HJK9810/2026-08-25/high/5644.java
import java.util.ArrayList;
import java.util.Scanner;

class Solution {
  private static final int[][] DIR = new int[][] {{0, 0}, {-1, 0}, {0, 1}, {1, 0}, {0, -1}};
  private int moveCount;

  private int[][] sortCharge(ArrayList<int[]> chargeCases) {
    int size = chargeCases.size();
    int[][] sortedCase = new int[size][2];

    for (int idx = 0; idx < size; idx++) {
      sortedCase[idx] = chargeCases.get(idx);
    }

    for (int end = size - 1; end >= 0; end--) {
      for (int index = 0; index < end; index++) {
        if (sortedCase[index][1] < sortedCase[index + 1][1]) {
          int[] temp = sortedCase[index + 1];
          sortedCase[index + 1] = sortedCase[index];
          sortedCase[index] = temp;
        }
      }
    }

    return sortedCase;
  }

  private int[] movePos(int[] current, int dirNum) {
    int dRow = DIR[dirNum][0];
    int dCol = DIR[dirNum][1];

    return new int[] {current[0] + dRow, current[1] + dCol};
  }

  private ArrayList<int[]>[][] initialBoard(ArrayList<int[]> inputCases) {
    ArrayList<int[]>[][] boards = new ArrayList[10][10];

    for (int row = 0; row < 10; row++) {
      for (int col = 0; col < 10; col++) {
        boards[row][col] = new ArrayList<>();
        boards[row][col].add(new int[] {0, 0});
      }
    }

    for (int idx = 0; idx < inputCases.size(); idx++) {
      int[] inputCase = inputCases.get(idx);
      int col = inputCase[0];
      int row = inputCase[1];
      int size = inputCase[2];
      int charge = inputCase[3];

      for (int r = row - size; r <= row + size; r++) {
        for (int c = col - size; c <= col + size; c++) {
          if (r >= 0 && r < 10 && c >= 0 && c < 10 && Math.abs(r - row) + Math.abs(c - col) <= size) {
            boards[r][c].add(new int[] {idx + 1, charge});
          }
        }
      }
    }

    return boards;
  }

  private int algorithm(ArrayList<int[]> inputCase, ArrayList<Integer> aMoveInput, ArrayList<Integer> bMoveInput) {
    ArrayList<int[]>[][] boards = initialBoard(inputCase);

    int totalCharge = 0;
    aMoveInput.add(0, 0);
    bMoveInput.add(0, 0);

    int[] aPos = new int[] {0, 0};
    int[] bPos = new int[] {9, 9};

    for (int move = 0; move < moveCount + 1; move++) {
      aPos = movePos(aPos, aMoveInput.get(move));
      bPos = movePos(bPos, bMoveInput.get(move));

      int[][] aCharge = sortCharge(boards[aPos[0]][aPos[1]]);
      int[][] bCharge = sortCharge(boards[bPos[0]][bPos[1]]);

      if (aCharge.length == 1 && aCharge.length == bCharge.length) continue;
      else if (aCharge[0][0] == bCharge[0][0]) {
        totalCharge += (aCharge[0][1] + Math.max(aCharge[1][1], bCharge[1][1]));
      } else {
        totalCharge += (aCharge[0][1] + bCharge[0][1]);
      }
    }

    return totalCharge;
  }

	public static void main(String args[]) throws Exception {
    Solution sol = new Solution();

		Scanner sc = new Scanner(System.in);
		int T = sc.nextInt();

		for(int test_case = 1; test_case <= T; test_case++) {
      sol.moveCount = sc.nextInt();
      int A = sc.nextInt();
      ArrayList<int[]> inputCase = new ArrayList<>();
      ArrayList<Integer> aMoveInput = new ArrayList<>();
      ArrayList<Integer> bMoveInput = new ArrayList<>();
      
      for (int idx = 0; idx < sol.moveCount; idx++) {
        aMoveInput.add(sc.nextInt());
      }
      for (int idx = 0; idx < sol.moveCount; idx++) {
        bMoveInput.add(sc.nextInt());
      }
      for (int idx = 0; idx < A; idx++) {
        inputCase.add(new int[] {sc.nextInt() - 1, sc.nextInt() - 1, sc.nextInt(), sc.nextInt()});
      }

      int result = sol.algorithm(inputCase, aMoveInput, bMoveInput);

      System.out.printf("#%d %d\n", test_case, result);
		}

    sc.close();
	}
}
